ipol_demo/modules/archive/archive.py

```diff
@@ -24,1013 +9,950 @@
+# TODO change this name
+# TODO No usage?
+@app.get("/executions_per_demo", status_code=200)
+def executions_per_demo():
+    """
+    return the number of executions per demo
+    """
+    conn = None
+    try:
+        conn = lite.connect(settings.database_file)
+        cursor_db = conn.cursor()
 
-        Archive.mkdir_p(self.database_dir)
-        Archive.mkdir_p(self.blobs_dir)
-        Archive.mkdir_p(self.blobs_thumbs_dir)
+        cursor_db.execute(
+            """
+            SELECT id_demo, COUNT(*)
+            FROM experiments
+            GROUP BY id_demo
+        """
+        )
+        data = {id: nb for id, nb in cursor_db.fetchall()}
+
+    except Exception as ex:
+        message = "Failure in stats function. Error: {}".format(ex)
+        print(message)
+        log.exception(message)
+    finally:
+        if conn is not None:
+            conn.close()
 
-        # Logs
-        try:
-            if not os.path.exists(self.logs_dir):
-                os.makedirs(self.logs_dir)
-            self.logger = self.init_logging()
-        except Exception as ex:
-            print("Failed to create log dir. Error: {}".format(ex))
+    return data
 
-        # Security: authorized IPs
-        self.authorized_patterns = self.read_authorized_patterns()
 
-        try:
-            self.database_name = cherrypy.config.get("database_name")
-        except Exception:
-            self.database_name = "archive.db"
+# TODO only used in archive tests
+@app.get("/demo_list", status_code=200)
+def demo_list():
+    """
+    Return the list of demos with at least one experiment
+    """
+    demo_list = []
+    try:
+        conn = lite.connect(settings.database_file)
+        cursor_db = conn.cursor()
+        cursor_db.execute(
+            """
+        SELECT DISTINCT id_demo FROM experiments"""
+        )
 
-        self.database_file = os.path.join(self.database_dir, self.database_name)
+        for row in cursor_db.fetchall():
+            demoid = row[0]
+            demo_list.append(demoid)
 
-        if not self.init_database():
-            sys.exit("Initialization of database failed. Check the logs.")
+        conn.close()
+    except Exception as ex:
+        message = "Failure in demo_list. Error = {}".format(ex)
+        log.exception(message)
+        raise HTTPException(status_code=404, detail=message)
+    return {"demo_list": demo_list}
 
-    def read_authorized_patterns(self):
-        """
-        Read from the IPs conf file
-        """
-        # Check if the config file exists
-        authorized_patterns_path = os.path.join(self.config_common_dir, "authorized_patterns.conf")
-        if not os.path.isfile(authorized_patterns_path):
-            self.error_log("read_authorized_patterns", "File {} doesn't exist".format(authorized_patterns_path))
-            return []
 
-        # Read config file
+@app.post("/experiment", status_code=201)
+def create_experiment(experiment: Experiment):
```

Comment:
We probably need to add return types as well https://fastapi.tiangolo.com/tutorial/response-model/.

Reply:
I was adding them for exposed endpoints, shall I do the same for helper functions and it's parameter types?

Reply:
It's particularly important for all endpoints, but it's best to do it for all python functions as it will allow to run static type checkers (mypy) in the CI in the future.

Reply:
I've added return types and parameter types where I could. I haven't found a way yet to set up a return type when a function returns two values.

Reply:
For example if you have a function:
```python
def add_blob(conn, blob_dict, copied_files_list: list):
        return id_blob, blob_name
```
the return type should probably be `tuple[int, str]`
So:
```python
def add_blob(conn, blob_dict, copied_files_list: list) -> tuple[int, str]:
```

For dict this is the same, you should specify the type of the key and value ; if the values can have different types in the dict, this is a good indication that it should be converted into a class.